### ISSUE TITLE
Charset support: Convert bytes to chars

### ISSUE THREAD

**Issue by ennru** [akka/alpakka#569]
Text based readers could need help with decoding incoming byte streams to characters according to a specified character set, as discussed in #502 "csv does not play well with framing".
Simply relying on `ByteString.decode` might fail in cases where a multi-byte character happens to be split across two stream elements of `ByteString`.

To support charsets properly we'd need to build something based on `nio.charset.CharsetDecoder` to provide a `Flow[ByteString, String]` or maybe a `Flow[ByteString, CharString]` where CharString would be some immutable wrapper of `nio.CharBuffer`.
Delivering Strings is simple of course, but it glosses over the fact that the Strings won't have proper framing by lines or other entities. Inventing `CharString`s makes this more obvious. 

With proper decoding in place text parsing stages as CSV could work on characters instead of bytes.

The `CharsetDecoder` API is very low-level and interacts tightly with NIO buffers.

Any thoughts on this idea?

**Comment by johanandren** [comment]
Sounds like quite a bit of work for something I don't quite see the concrete problem with, that you would use multi-byte-charsets and by accident match the separator byte in byte two or later of a multi-byte character?

Or was it to correctly set an upper limit in characters rather than bytes? I don't think that either warrants the work to do this. After all having memory filling up is about running out of bytes, not characters.

**Comment by ennru** [comment]
Maybe a misleading issue reference - this idea is not about limiting the line length for CSV.

It is about supporting other than "mostly single byte" character sets (ASCII, UTF-8), as the CSV parse does right now (e.g. UTF-16 is unsupported). The CSV parser relies on single-byte delimiter chars as it is quite hard to decode multi-byte characters properly.

Other text-based formats do have the same problem (even though I've never seen JSON encoded in UTF-16 in the wild):
```
      val fut = Source
        .single(ByteString("{a,b}".getBytes(StandardCharsets.UTF_16BE)))
        .via(JsonFraming.objectScanner(1024))
        .runWith(Sink.seq)
```
fails with
`akka.stream.scaladsl.Framing$FramingException, with message: Invalid JSON encountered at position [0] of [ByteString(0, 123, 0, 97, 0, 44, 0, 98, 0, 125)]`

**Comment by johanandren** [comment]
Is having a multi byte character as delimiter in CSV really something people do? Would be good to see someone actually having that problem tell about it so it isn't a fabricated problem. :)

Same goes for the JsonFraming, even if UTF-16 or UTF-32 is valid, I have neither seen it in the wild nor heard anyone ask about it. But maybe I'm just a lucky european.

**Comment by ennru** [comment]
Same here, but I think other encodings are used much more on non-JVM systems. From the integration point of view of Alpakka it demands support for streamed character decoding, or maybe re-coding to JVM standard UTF-8.

Maybe @voltir has other concrete examples? (He started the byte vs char idea in #502.)

**Comment by Voltir** [comment]
Yea, so my issue came about due to a client wanting to use `Ç` as the delimiter. Now, the intent of using that character was ostensibly that it is easy to split on - no need to deal with csv escaping if you have a delimiter that never shows up in data.

But I was attempting to make use of the alpakka csv parser anyway, and that character is particularly annoying - it is a single byte character in the source encoding (windows-1252) but is actually a 3 byte character in UTF8 - in short, I simply cannot use alpakka's csv parser on the strange delimiter. 

My solution is simple - just split on the character and ignore the intricacies of csv's altogether . However, I still think there is a fundamental problem with attempting to do csv parsing at they byte level - it really is kind of nonsensical to attempt to interpret csvs out of raw bytes -- csv's are built out of text files and text files have an encoding. I think solving the encoding issue, and then building text based flows on top of that will be very valuable to the ecosystem as a whole, and will radically simplify the csv code in particular.

**Comment by raboof** [comment]
While it would make sense from a conceptual point of view, encoding/decoding can be memory- and cpu-intensive task, so for the common case of splitting on lower-ASCII characters in an UTF-8 stream postponing decoding (or, for some use cases, avoiding it entirely) can be a great performance advantage.

If the delimiter is a single byte character in your source encoding, wouldn't it be possible to split on that character in the raw windows-1252 bytes? Or do you receive the file after it has been converted to UTF-8?

**Comment by Voltir** [comment]
I do not need my csv parsing library to make "performance optimization" choices for me, especially at the cost of correctness. The original problem has been solved long ago, and yes we manage all the byte encoding issues ourselves.

There are a ton of uses for being able to deal with text encoding in a streaming environment, and I very much want alpakka to be a library of building blocks - let me, the end library user, deal with the performance constraints of my application, I really just want the tools to build it.

**Comment by ennru** [comment]
Just had a closer look at PR #320 which contains a `CharDecoderFlow`. Let's see where it goes.

**Comment by jypma** [comment]
Oh feel free to just steal those flows from #320 into a separete PR.  #320 may take a while, given its size ;-)

**Comment by ennru** [comment]
I've started to look into this again. Based on @jypma 's Character Decoding/Encoding implementations I'm implementing a *Transcoding* stage which offers a `Flow[ByteString, ByteString, _]` to move character data from one encoding to another without creating `String`s in between.
Transcoding will support any JVM supported Character encoding to all text-consuming stages that today expect `ByteString`s with UTF-8 content.

**Comment by ennru** [comment]
Fixed with #1011
